### ISSUE TITLE
continueOn: failed: true doesn't work as expected for "pod deleted" steps

### ISSUE THREAD

**Issue by chtcvl** [argoproj/argo-workflows#5738]
## Summary 

What happened/what you expected to happen?
What happened:
* Some pods get `pod deleted` error message, even though the workload succeeds and main.log successfully stored to artifacts. This happens in a nested workflow and the next step of the nested workflow never happens even though there is:
```
        continueOn:
          failed: true
```
Expected:
1. Not having `pod deleted` on pods which actually succeed (can see on the logs exitCode is set to 0 successfully, artifacts stored)
2. Allow next step to proceed (since `continueOn: failed: true`) despite previous step erring with `pod deleted` 

## Diagnostics

What Kubernetes provider are you using? 
GKE
What version of Argo Workflows are you running? 
v3.0.1
```yaml

apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: redacted-to-l2
spec:
  arguments:
    parameters:
      - name: date
      - name: inp
      - name: query-limit
        value: "99999"
  entrypoint: loop-param-result
  volumes:
    - name: workdir
      emptyDir: {}
  templates:
  - name: loop-param-result
    inputs:
      parameters:
        - name: date
        - name: query-limit
    steps:
    - - name: generate
        template: gen-params
        arguments:
          parameters:
          - name: date
            value: "{{inputs.parameters.date}}"
          - name: query-limit
            value: "{{inputs.parameters.query-limit}}"
    # Iterate over the list of params generated by the generate step above
    - - name: workload-wrapper
        template: workload-wrapper
        arguments:
          parameters:
          - name: inp
            value: "{{item}}"
          - name: date
            value: "{{inputs.parameters.date}}"
        withParam: "{{steps.generate.outputs.result}}"
        continueOn:
          failed: true
    # Finally, aggregate and print all failed runs from previous steps
    - - name: print-failed-results
        template: print-failed-results

  #### Generate a list of parameters in JSON format
  - name: gen-params
    inputs:
      parameters:
        - name: date
        - name: query-limit
    initContainers:
      - image:  gcr.io/google.com/cloudsdktool/cloud-sdk:326.0.0-alpine
        name: workload-identity-initcontainer
        command:
        - '/bin/bash'
        - '-c'
        - |
          curl -s -H 'Metadata-Flavor: Google' 'http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token' --retry 30 --retry-connrefused --retry-max-time 30 > /dev/null || exit 1
    container:
      image: registry:base10
      imagePullPolicy: "IfNotPresent"
      command: ["python3", "-W", "ignore", "/src/genparams.py"]
      args:
        - '--date={{inputs.parameters.date}}'
        - '--limit={{inputs.parameters.query-limit}}'
      resources:
        limits:
          memory: 1100Mi
          cpu: 1100m
        requests:
          memory: 1Gi
          cpu: "1"
  - name: workload-wrapper
    inputs:
      parameters:
        - name: inp
        - name: date
    steps:
    - - name: workload
        template: workload
        arguments:
          parameters:
          - name: inp
            value: "{{inputs.parameters.inp}}"
          - name: date
            value: "{{inputs.parameters.date}}"
        continueOn:
          failed: true
    - - name: evaluate-result
        template: evaluate-result
        arguments:
          parameters:
          - name: run_id
            value: "{{inputs.parameters.inp}}-{{inputs.parameters.date}}"
          - name: res
            value: "{{steps.workload.status}}"
          - name: log_path
            value: "{{steps.workload.outputs.parameters.log_path}}"
  #### actual workload
  - name: workload
    inputs:
      parameters:
        - name: inp
        - name: date
    initContainers:
      - image:  gcr.io/google.com/cloudsdktool/cloud-sdk:326.0.0-alpine
        name: workload-identity-initcontainer
        command:
        - '/bin/bash'
        - '-c'
        - |
          curl -s -H 'Metadata-Flavor: Google' 'http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token' --retry 30 --retry-connrefused --retry-max-time 30 > /dev/null || exit 1
    container:
      image: registry:base10
      imagePullPolicy: "IfNotPresent"
      command: ["python3","/src/datal2prep.py"]
      args:
        - "--inp={{inputs.parameters.inp}}"
        - "--date={{inputs.parameters.date}}"
        - "--dest=gs://bucket/dir”
      resources:
        limits:
          memory: 1100Mi
          cpu: 1100m
        requests:
          memory: 1Gi
          cpu: "1"
    outputs:
      parameters:
        - name: run_id
          value: "{{inputs.parameters.inp}}-{{inputs.parameters.date}}"
        - name: log_path
          value: "artifacts/{{workflow.namespace}}/{{workflow.creationTimestamp.Y}}-{{workflow.creationTimestamp.m}}-{{workflow.creationTimestamp.d}}-{{pod.name}}/main.log"
  ### return workload exit code if it failed
  - name: evaluate-result
    inputs:
      parameters:
      - name: run_id
      - name: res
      - name: log_path
    nodeSelector:
      nodeclass: ephemeral
    script:
      image: registry:base10
      command: [python3]
      source: |
        import json
        import os
        import sys
        [redacted]
      volumeMounts:
      - name: workdir
        mountPath: /mnt/out
    outputs:
      artifacts:
        - name: part
          path: /mnt/out/part.json
          archive:
            none: { }
          gcs:
            key: "{{workflow.name}}/results/{{inputs.parameters.run_id}}.json"
  # print only runs where result is failed from a list of outputs. 
  - name: print-failed-results
    inputs:
      artifacts:
        - name: results
          path: /mnt/in
          gcs:
            bucket: bucket-name
            key: "{{workflow.name}}/results"
    script:
      image: redacted:py38
      command: [python3]
      source: |
        import json
        import os
        import sys
        [redacted]
      volumeMounts:
      - name: workdir
        mountPath: /mnt
    outputs:
      artifacts:
        - name: total
          path: /mnt/out/total.json
          archive:
            none: { }
          gcs:
            bucket: bucket-name
            key: "{{workflow.name}}/failed_runs.json"
  ttlStrategy:
    secondsAfterCompletion: 300
    secondsAfterSuccess: 300
    secondsAfterFailure: 300
  podGC:
    #strategy: OnPodCompletion
    strategy: OnWorkflowCompletion
```
```
time="2021-04-22T12:13:11.289Z" level=info msg="Updating node redacted-to-l2-l4hqm-953958221 status Pending -> Running" namespace=workflow-ns workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:14:20.054Z" level=info msg="Setting node redacted-to-l2-l4hqm-953958221 outputs" namespace=workflow-ns workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:14:47.710Z" level=debug msg="Executing node redacted-to-l2-l4hqm[1].workload-wrapper(3301:re|da|6357|)[0].workload with container template: &Template{Name:workload,Inputs:Inputs{Parameters:[]Parameter{Parameter{Name:sym,Default:nil,Value:*re|da|6357|,ValueFrom:nil,GlobalName:,Enum:[],},Parameter{Name:date,Default:nil,Value:*today,ValueFrom:nil,GlobalName:,Enum:[],},},Artifacts:[]Artifact{},},Outputs:Outputs{Parameters:[]Parameter{Parameter{Name:run_id,Default:nil,Value:*re|da|6357|-today,ValueFrom:nil,GlobalName:,Enum:[],},Parameter{Name:log_path,Default:nil,Value:*artifacts/workflow-ns/2021-04-22-redacted-to-l2-l4hqm-953958221/main.log,ValueFrom:nil,GlobalName:,Enum:[],},},Artifacts:[]Artifact{},Result:nil,ExitCode:nil,},NodeSelector:map[string]string{},Affinity:nil,Metadata:Metadata{Annotations:map[string]string{},Labels:map[string]string{},},Daemon:nil,Steps:[]ParallelSteps{},Container:&v1.Container{Name:,Image:us-central1-docker.pkg.dev/addr-mgmt/dkr/addr-job-addr-market-data:base10,Command:[python3 /src/addrl2prep.py],Args:[--sym=re|da|6357| --date=today --dest=gs://addr-shared-cloud-storage/addrL2],WorkingDir:,Ports:[]ContainerPort{},Env:[]EnvVar{EnvVar{Name:GCE_METADATA_TIMEOUT,Value:30s,ValueFrom:nil,},},Resources:ResourceRequirements{Limits:ResourceList{cpu: {{1100 -3} {<nil>} 1100m DecimalSI},memory: {{1153433600 0} {<nil>} 1100Mi BinarySI},},Requests:ResourceList{cpu: {{1 0} {<nil>} 1 DecimalSI},memory: {{1073741824 0} {<nil>} 1Gi BinarySI},},},VolumeMounts:[]VolumeMount{},LivenessProbe:nil,ReadinessProbe:nil,Lifecycle:nil,TerminationMessagePath:,ImagePullPolicy:IfNotPresent,SecurityContext:nil,Stdin:false,StdinOnce:false,TTY:false,EnvFrom:[]EnvFromSource{},TerminationMessagePolicy:,VolumeDevices:[]VolumeDevice{},StartupProbe:nil,},Script:nil,Resource:nil,DAG:nil,Suspend:nil,Volumes:[]Volume{},InitContainers:[]UserContainer{UserContainer{Container:{workload-identity-initcontainer gcr.io/google.com/cloudsdktool/cloud-sdk:326.0.0-alpine [/bin/bash -c curl -s -H 'Metadata-Flavor: Google' 'http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token' --retry 30 --retry-connrefused --retry-max-time 30 > /dev/null || exit 1\n] []  [] [] [] {map[] map[]} [] [] nil nil nil nil    nil false false false},MirrorVolumeMounts:nil,},},Sidecars:[]UserContainer{},ArchiveLocation:nil,ActiveDeadlineSeconds:<nil>,RetryStrategy:nil,Parallelism:nil,Tolerations:[]Toleration{},SchedulerName:,PriorityClassName:,Priority:nil,ServiceAccountName:,HostAliases:[]HostAlias{},SecurityContext:nil,PodSpecPatch:,AutomounaddrrviceAccountToken:nil,Executor:nil,Metrics:nil,Synchronization:nil,Memoize:nil,Timeout:,}\n" namespace=workflow-ns workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:14:47.710Z" level=debug msg="Skipped pod redacted-to-l2-l4hqm[1].workload-wrapper(3301:re|da|6357|)[0].workload (redacted-to-l2-l4hqm-953958221) creation: already exists" namespace=workflow-ns podPhase=Succeeded workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:14:56.073Z" level=info msg="node redacted-to-l2-l4hqm-953958221 phase Running -> Error" namespace=workflow-ns workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:14:56.073Z" level=info msg="node redacted-to-l2-l4hqm-953958221 message: pod deleted" namespace=workflow-ns workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:14:56.073Z" level=info msg="node redacted-to-l2-l4hqm-953958221 finished: 2021-04-22 12:14:56.073139748 +0000 UTC" namespace=workflow-ns workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:15:15.565Z" level=info msg="Step group node redacted-to-l2-l4hqm-3332645986 deemed failed: child 'redacted-to-l2-l4hqm-953958221' failed" namespace=workflow-ns workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:15:15.565Z" level=info msg="node redacted-to-l2-l4hqm-3332645986 message: child 'redacted-to-l2-l4hqm-953958221' failed" namespace=workflow-ns workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:15:15.565Z" level=info msg="step group redacted-to-l2-l4hqm-3332645986 was unsuccessful: child 'redacted-to-l2-l4hqm-953958221' failed" namespace=workflow-ns workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:15:15.565Z" level=info msg="Outbound nodes of redacted-to-l2-l4hqm-953958221 is [redacted-to-l2-l4hqm-953958221]" namespace=workflow-ns workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:15:15.565Z" level=info msg="Outbound nodes of redacted-to-l2-l4hqm-4147985912 is [redacted-to-l2-l4hqm-953958221]" namespace=workflow-ns workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:15:15.565Z" level=info msg="node redacted-to-l2-l4hqm-4147985912 message: child 'redacted-to-l2-l4hqm-953958221' failed" namespace=workflow-ns workflow=redacted-to-l2-l4hqm
time="2021-04-22T12:15:15.566Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"workflow-ns\", Name:\"redacted-to-l2-l4hqm\", UID:\"020ed8cd-beac-4c48-8a03-49e4f91ae126\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"128561118\", FieldPath:\"\"}): type: 'Warning' reason: 'WorkflowNodeFailed' Failed node redacted-to-l2-l4hqm[1].workload-wrapper(3301:re|da|6357|)[0]: child 'redacted-to-l2-l4hqm-953958221' failed"
time="2021-04-22T12:15:15.566Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"workflow-ns\", Name:\"redacted-to-l2-l4hqm\", UID:\"020ed8cd-beac-4c48-8a03-49e4f91ae126\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"128561118\", FieldPath:\"\"}): type: 'Warning' reason: 'WorkflowNodeFailed' Failed node redacted-to-l2-l4hqm[1].workload-wrapper(3301:re|da|6357|): child 'redacted-to-l2-l4hqm-953958221' failed"

```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by terrytangyuan** [comment]
I believe this error should stop appearing after https://github.com/argoproj/argo-workflows/pull/4808 got merged (related issue https://github.com/argoproj/argo-workflows/issues/4798) but I'll defer this to @alexec to double check.

**Comment by chtcvl** [comment]
There are no events regarding pod eviction in events. In cluster audit log, Pod is deleted by `kube-system:pod-garbage-collector` service account.

**Comment by alexec** [comment]
I think this will be fixed in v3.0.4
